### PR TITLE
Fix drag-and-drop sorting on Firefox Linux (fixes #19564)

### LINKED ISSUE #19564: Sort children is broken in Firefox on Linux
### Which Umbraco version are you using?

main (V16@50272f74)

### Bug summary

Content cannot be sorted. 

### Specifics

The drag and drop in the "Sort Children" does not work on Firefox 139.0.4

Details: 
Browser: Firerfox 139.0.4  
OS: Kubuntu 24.04.2

https://github.com/user-attachments/assets/de54bad6-d3ad-4fc5-bf69-e3a93b26ec44

### Steps to reproduce

1. Create multiple pieces of content
2. Try to change the sort order

### Expected result / actual result

Expected: Drag and drop moves the content
Atual: Drag and drop does not move the content

### PR DESCRIPTION
**Pull request by nul800sebastiaan**
Firefox on Linux fires mouseup events and reports buttons === 0 during active drag operations, which prematurely cleared the dropSorter state.

Check if UmbSorterController.activeItem exists before ending drag to prevent premature cleanup during active drag operations.

Also call preventDefault() at start of dragover handler to ensure Firefox recognizes the drop zone.

Fixes sorting in content tree, doctype editor, and block lists.

### Prerequisites

- [x] I have added steps to test this contribution in the description below

Fixes #19564

## Root Cause
Firefox on Linux has two platform-specific quirks with drag-and-drop:

1. **`mouseup` events fire during active drag operations** - Even though the mouse button is still held down, Firefox on Linux fires `mouseup` events during drag.

2. **`mousemove` events report `buttons === 0` during drag** - The `MouseEvent.buttons` property incorrectly reports 0 (no buttons pressed) even during active drag.

These quirks caused the sorter to prematurely call `#handleMoveEnd()`, which cleared the `dropSorter` state before any `dragover` events could process the drag.

### Changes to `sorter.controller.ts`

**1. Check for active drag in `#handleMouseUp`** (line ~916)
```ts
#handleMouseUp = (event?: MouseEvent) => {
    // Firefox on Linux bug: mouseup can fire during active drag, so don't end if actively dragging
    if (!UmbSorterController.activeItem) {
        this.#handleMoveEnd();
    }
};
```

**2. Check for active drag in `#handleMouseMove`** (line ~902)
```ts
#handleMouseMove = (event: MouseEvent) => {
    // Firefox on Linux bug: buttons can be 0 during active drag, so don't end if actively dragging
    if (event.buttons === 0 && !UmbSorterController.activeItem) {
        this.#handleMoveEnd();
    }
};
```

**3. Handle already-set dropSorter in `#obtainIncomingItem`** (line ~487)
```ts
async #obtainIncomingItem(e: DragEvent) {
    // If dropSorter is already set and it's this sorter, we're already handling this drag
    if (UmbSorterController.dropSorter && UmbSorterController.dropSorter.identifier === this.identifier) {
        return true;
    }
    // ... rest of method
}
```

**4. Call `preventDefault()` at start of `#itemDraggedOver`** (line ~559)
```ts
#itemDraggedOver = async (e: DragEvent) => {
    // Firefox on Linux requires preventDefault to be called on dragover for drops to work
    e.preventDefault();
    // ... rest of method
}
```

## Why This Works

The fix prevents premature cleanup by checking if `UmbSorterController.activeItem` exists before calling `#handleMoveEnd()`. This ensures:

1. `dropSorter` remains set throughout the drag operation
2. `dragover` events can properly handle the drag
3. `#handleDragMove` gets called to reorder items
4. Dropping completes successfully

The `activeItem` property is set during `dragstart` and only cleared at the end of a successful drag or explicit cancellation, making it a reliable indicator of whether a drag is in progress.

## Testing

To verify the fix works:

1. Open Firefox on Linux
2. Navigate to any Umbraco interface with sortable items:
   - Content tree: Sort child nodes
   - Document Type editor: Sort properties
   - Block List: Use Sort mode
3. Drag an item - other items should move out of the way
4. Drop the item - it should stay in the new position

## Related Issues

- Line 798: Existing Firefox workaround for ShadowDOM drag images: `transform: translateZ(0)`
- Commit 84fecd3521: Fixed Linux CTRL+Click platform detection

## Browser Compatibility Notes

This fix is specific to Firefox on Linux and does not affect behavior on other browsers or platforms. The checks for `activeItem` are safe on all platforms as they simply verify that a drag is actually in progress before ending it.

### CODE REVIEW

**Comment by nul800sebastiaan** [comment]
Manually tested on Linux:
- Firefox
- Chrome
- Vivaldi 

Tested:sorting nodes, sort mode in the document type editor, drag & drop in block list and block grid (both in editing mode and in sort mode).

Manually tested on Windows:
- Firefox
- Chrome

Same areas tested as above.

File: src/Umbraco.Web.UI.Client/src/packages/core/sorter/sorter.controller.ts [hunk 1]
```diff
@@ -844,17 +851,24 @@ export class UmbSorterController<T, ElementType extends HTMLElement = HTMLElemen
...
 	 * Listen to mouse move, to check if the mouse is still down.
 	 * This event does not happen while dragging, so its a indication that the drag is over.
+	 * NOTE: On Firefox Linux, mousemove can fire during drag with buttons === 0, so we need to be careful.
 	 */
 	#handleMouseMove = (event: MouseEvent) => {
 		// buttons should reprensent which buttons are held, and 0 => represents no button is pressed. [NL]
-		if (event.buttons === 0) {
+		// Firefox on Linux bug: buttons can be 0 during active drag, so don't end if actively dragging
+		if (event.buttons === 0 && !UmbSorterController.activeItem) {
+			// Only end the move if there's no active drag item
```
[the remark this answers is not in the document]
Comment (nul800sebastiaan) [reply src/Umbraco.Web.UI.Client/src/packages/core/sorter/sorter.controller.ts:857]: Added tests in `src/Umbraco.Web.UI.Client/src/packages/core/sorter/sorter.controller.test.ts`.

**Comment by nul800sebastiaan** [comment]
The build errors are unrelated, possibly a problem at NuGet for a while? 

image

**Comment by AndyButland** [comment]
The build error does seem unrelated - and I fired them off again and the failing check passed this time.

This looks good to me and given you've already tested on all those platforms - for which, thanks! - I would say this is safe to merge (i.e. it fixes the issue on Firefox/Linux but doesn't cause problems on other platforms).

@iOvergaard @nielslyngsoe - could one have a quick look over the code in case you have any concerns or suggestions please?

File: src/Umbraco.Web.UI.Client/src/packages/core/sorter/sorter.controller.ts [hunk 2]
```diff
@@ -844,17 +848,24 @@ export class UmbSorterController<T, ElementType extends HTMLElement = HTMLElemen
...
+			// If activeItem exists, we're in the middle of a drag and shouldn't end it
 			this.#handleMoveEnd();
 		}
 	};
 
 	// eslint-disable-next-line @typescript-eslint/no-unused-vars
 	#handleMouseUp = (event?: MouseEvent) => {
-		this.#handleMoveEnd();
+		// Firefox on Linux bug: mouseup can fire during active drag, so don't end if actively dragging
+		if (!UmbSorterController.activeItem) {
```
Comment (nielslyngsoe) [inline src/Umbraco.Web.UI.Client/src/packages/core/sorter/sorter.controller.ts:866]: This means, if there is an active item, then we will not handle the end?

That seems to me like we just stopped cleaning up? So, not sure if this is a solution, I will do some investigations to see what we could do to not react to these event coming while dragging.

File: src/Umbraco.Web.UI.Client/src/packages/core/sorter/sorter.controller.test.ts [hunk 3]
```diff
@@ -209,4 +209,94 @@ describe('UmbSorterController', () => {
 			expect(element.sorter.getItem('5')).to.be.undefined;
 		});
 	});
+
+	describe('Firefox Linux edge cases (fixes #19564)', () => {
```
Comment (nielslyngsoe) [inline src/Umbraco.Web.UI.Client/src/packages/core/sorter/sorter.controller.test.ts:213]: Does this test prove a problem when running on FireFox in Linux?

Cause on my laptop this does not prove anything. If I revert the code back to what it was before, this does not fail. — I would be skeptical about adding this test if it does not help us prevent this issue in the future.

File: src/Umbraco.Web.UI.Client/src/packages/core/sorter/sorter.controller.ts (hunk 2 above)
Comment (nielslyngsoe) [reply src/Umbraco.Web.UI.Client/src/packages/core/sorter/sorter.controller.ts:866]: I looked into this. The reason why we observe mouse-up and mouse-move is to detect if the mouse comes back without dragging.

The problem is we do not know if an item has been dropped outside this viewport. It could have been dragged to another tab or other scenarios.
— We cannot know because of Web-Components and the recommendation of re-rendering them when re-ordering them. That means the original element lost the connection to the DOM.

Therefore, we detect if the mouse comes back with nothing. and concludes that it means the drag session has ended. This is very important cause otherwise we never end the drag-state.

You can try this your self by dragging an item outside the view port and drop it there.

So the solution to this would be to see if there is other ways to detect this in FireFox on Linux.

**Comment by nielslyngsoe** [comment]
@nul800sebastiaan I have marked this PR as draft as this solution didn't seem to fix anything, it just seems to turn of logic, which in a happy path was alright, but when dropping items outside the viewport, the problems surfaced.

So based on your inputs i came up with another solution, could you verify that it works:
https://github.com/umbraco/Umbraco-CMS/pull/21664

Thanks

**Comment by nul800sebastiaan** [comment]
Ha! I am totally embarrassed that I just let my friend Uncle Claude generate a test for me, thinking it would help the PR but of course it doesn't prove a damned thing.. sorry about that!
